### DISCUSSION TITLE
array of variables

### DISCUSSION THREAD

**Discussion by mortezaaesmaeilpour** [idaholab/moose#21234] [Q&A General] [answered]
Hi all

The mass, momentum, and energy equations are solved (and coupled) in my MOOSE-based application to simulate multiphase flows in wellbores. Now, I want to transport different components in the simulation environment (like CO2, N2, NaCl, ...). a new main variable should be defined for the concentration of each of these new components in the mixture. The new kernels are like each other except the off-diagonal jacobian part ( concentrations of other components). 

1. My first question is: what is the best way to define these new kernels and variables (action, array kernel, ....). Could you please address an example in MOOSE?
2. My second question is: How can I compute or define off-diagonal jacobians in mass, momentum, and energy kernels against these new sets of variables (concentration)?

Kind regards
Morteza

**Comment by GiudGiud** [comment, marked as answer]
Hello

The best way is to use array variables and array kernels if there are quite a few components. This will reduce the number of objects needed to tackle the equations and possibly will reduce the computational time by reducing duplicate calculations. 

Well is there off-diagonal jacobian terms for mass momentum and energy? Are the concentrations in co2 etc influencing the material properties, like density or thermal conductivity? If so then I think the cleaner way to do this is to use automatic differentiation. The derivatives will automatically be carried forward from the concentrations to those equations

Just out of curiosity, what numerical discretization did you use for this problem? FEM ? FV? What scheme?

Guillaume
